sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -537,12 +537,19 @@ object CollapseProject extends Rule[LogicalPlan] {
 }
 
 /**
- * Combines adjacent [[Repartition]] operators by keeping only the last one.
+ * Combines adjacent [[Repartition]] and [[RepartitionByExpression]] operator combinations
+ * by keeping only the one.
  */
 object CollapseRepartition extends Rule[LogicalPlan] {
   def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
     case Repartition(numPartitions, shuffle, Repartition(_, _, child)) =>
       Repartition(numPartitions, shuffle, child)
+    case Repartition(numPartitions, _, r: RepartitionByExpression) =>
+      r.copy(numPartitions = Some(numPartitions))
+    case RepartitionByExpression(exprs, Repartition(_, _, child), numPartitions) =>
+      RepartitionByExpression(exprs, child, numPartitions)
+    case RepartitionByExpression(exprs, RepartitionByExpression(_, child, _), numPartitions) =>
+      RepartitionByExpression(exprs, child, numPartitions)
```

Comment:
Actually, it aims the followings SQLs. It's a kind of query organizing clause in SQL, e.g., ORDER BY, CLUSTER BY, DISTRIBUTE BY, SORT BY, LIMIT. Actually, the risk what you concern is about `exprs` depencency, right?
**Before**
```scala
scala> sql("(select * from values (1,'a'),(2,'b') T(a,b) distribute by a) distribute by b").explain
== Physical Plan ==
Exchange hashpartitioning(b#3, 200)
+- Exchange hashpartitioning(a#2, 200)
   +- LocalTableScan [a#2, b#3]
```

**After**
```scala
scala> sql("(select * from values (1,'a'),(2,'b') T(a,b) distribute by a) distribute by b").explain
== Physical Plan ==
Exchange hashpartitioning(b#1, 200)
+- LocalTableScan [a#0, b#1]
```